### PR TITLE
[JENKINS-52888] Add option to use AWS S3 accelerated endpoint

### PR DESCRIPTION
**Pull request by vblack**
[JENKINS-52888](https://issues.jenkins-ci.org/browse/JENKINS-52888)

Add option which allows to use S3 accelerated endpoint to speed-up
artifacts download from S3.

Signed-off-by: Kostiantyn Luzan <vblack2006@gmail.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -90,6 +90,7 @@ the same configuration page.
 * S3 Bucket Name: Name of the S3 Bucket to use to store artifacts.
 * S3 Bucket Region: Region to use to generate the URLs to get/put artifacts, by default it is autodetected.
 * Base Prefix: Prefix to use for files and folders inside the S3 Bucket, if the prefix is a folder should be end with `/`.
+* Use Accelerated Endpoint: Use accelerated endpoint for the S3 Bucket.
```
Comment (carlossg) [inline README.md]: I would link to https://docs.aws.amazon.com/AmazonS3/latest/dev/transfer-acceleration.html to explain what acceleration means

File: README.md [hunk 2]
```diff
@@ -230,6 +231,7 @@ Add to your `~/.m2/settings.xml`:
                 <AWS_REGION>…</AWS_REGION>
                 <S3_BUCKET>…</S3_BUCKET>
                 <S3_DIR>…/</S3_DIR>
+                <S3_ACCELERATED>…/</S3_ACCELERATED>
```
Comment (carlossg) [inline README.md]: `<S3_ACCELERATED>false</S3_ACCELERATED>`

File: sample-scripts/README.md [hunk 3]
```diff
@@ -1,2 +1,3 @@
 Example Pipeline scripts exercising various aspects of artifact storage.
-Will generally need `S3_BUCKET` and `S3_DIR` variables defined somewhere.
+Will generally need `S3_BUCKET`, `S3_DIR` and `S3_ACCELERATED` variables
```
Comment (carlossg) [inline sample-scripts/README.md]: `S3_ACCELERATED` is not needed, defaults to false

File: src/main/resources/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig/help-useAcceleratedEndpoint.html [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<div>This parameter is controll usage of accelerated endpoint,
```
Comment (carlossg) [inline src/main/resources/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig/help-useAcceleratedEndpoint.html]: should have a better description of what "accelerated endpoint" is, ie.

> Transfer Acceleration takes advantage of Amazon CloudFront’s globally distributed edge locations. As the data arrives at an edge location, data is routed to Amazon S3 over an optimized network path.
> When using Transfer Acceleration, additional data transfer charges may apply.

File: src/test/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3AbstractTest.java [hunk 5]
```diff
@@ -57,13 +57,15 @@
     private static final String S3_BUCKET = System.getenv("S3_BUCKET");
     protected static final String S3_DIR = System.getenv("S3_DIR");
     private static final String S3_REGION = System.getenv("S3_REGION");
+    private static final boolean S3_ACCELERATED = ((System.getenv("S3_ACCELERATED") == null) ? false : (System.getenv("S3_ACCELERATED").equals("true")));
```
Comment (carlossg) [inline src/test/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3AbstractTest.java]: A shorter way
`"true".equalsIgnoreCase(System.getenv("S3_ACCELERATED"));`

File: src/test/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3AbstractTest.java [hunk 6]
```diff
@@ -57,13 +57,15 @@
...
 
     protected BlobStoreProvider provider;
 
     @BeforeClass
     public static void live() {
         assumeThat("define $S3_BUCKET as explained in README", S3_BUCKET, notNullValue());
         assumeThat("define $S3_DIR as explained in README", S3_DIR, notNullValue());
+        assumeThat("define $S3_ACCELERATED as explained in README", S3_ACCELERATED, notNullValue());
```
Comment (carlossg) [inline src/test/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3AbstractTest.java]: S3_ACCELERATED  shouldn't be required

File: src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig.java [hunk 7]
```diff
@@ -216,7 +238,7 @@ public Bucket createS3Bucket(String name) throws IOException {
     }
 
     @RequirePOST
-    public FormValidation doCreateS3Bucket(@QueryParameter String container, @QueryParameter String prefix) {
+    public FormValidation doCreateS3Bucket(@QueryParameter String container, @QueryParameter String prefix, @QueryParameter Boolean acceleratedEndpoint) {
```
Comment (carlossg) [inline src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig.java]: no need to pass `acceleratedEndpoint` if it's not going to be used

File: src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig.java [hunk 8]
```diff
@@ -229,11 +251,11 @@ public FormValidation doCreateS3Bucket(@QueryParameter String container, @QueryP
     }
 
     @RequirePOST
-    public FormValidation doValidateS3BucketConfig(@QueryParameter String container, @QueryParameter String prefix) {
+    public FormValidation doValidateS3BucketConfig(@QueryParameter String container, @QueryParameter String prefix, @QueryParameter Boolean acceleratedEndpoint) {
```
Comment (carlossg) [inline src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig.java]: should deprecate the previous public method and add this as a new one

**Comment by vblack** [comment]
I'm not sure which Jira you are talking about, and where to track new feature.

File: README.md (hunk 1 above)
Comment (vblack) [reply README.md]: done

File: src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig.java (hunk 8 above)
Comment (carlossg) [reply src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig.java]: `@Deprecated`

**Comment by jglick** [comment]
> Please do not rebase changes until review is completed

There is no actual need to rebase changes in a PR at all. You can simply add follow-up commits to the same branch.

File: src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig.java [hunk 9]
```diff
@@ -79,6 +79,11 @@
     private String prefix;
     @Deprecated private transient String region, credentialsId;
 
+    /**
+     * Flag to use accelerated endpoint.
+     */
+    private Boolean acceleratedEndpoint;
```
Comment (jglick) [inline src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig.java]: `boolean` not `Boolean`.

Where are you getting this name? The docs do not mention anything about an “accelerated endpoint”. They refer to “transfer acceleration” and the APIs refer to “accelerate” or “accelerate mode”. We should use consistent terminology, both in the code and UI.

File: src/main/resources/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig/help-useAcceleratedEndpoint.html [hunk 10]
```diff
@@ -0,0 +1,4 @@
+<div>This parameter is controling usage of Amazon S3 Transfer Acceleration.
```
Comment (jglick) [inline src/main/resources/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig/help-useAcceleratedEndpoint.html]: typo

File: src/main/resources/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig/help-useAcceleratedEndpoint.html [hunk 11]
```diff
@@ -0,0 +1,4 @@
...
+Amazon S3 Transfer Acceleration enables fast, easy, and secure transfers of files over long distances between your client and an S3 bucket. Transfer Acceleration takes advantage of Amazon CloudFront’s globally distributed edge locations. As the data arrives at an edge location, data is routed to Amazon S3 over an optimized network path.
+<a href="https://docs.aws.amazon.com/AmazonS3/latest/dev/transfer-acceleration.html">https://docs.aws.amazon.com/AmazonS3/latest/dev/transfer-acceleration.html</a>
```
Comment (jglick) [inline src/main/resources/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig/help-useAcceleratedEndpoint.html]: The body of the link should be the page title or similar, not a copy of the URL.

File: src/test/java/io/jenkins/plugins/artifact_manager_jclouds/MockBlobStore.java [hunk 12]
```diff
@@ -67,6 +67,11 @@ public String getPrefix() {
         return "";
     }
 
+    @Override
+    public Boolean getAcceleratedEndpoint() {
```
Comment (jglick) [inline src/test/java/io/jenkins/plugins/artifact_manager_jclouds/MockBlobStore.java]: delete

**Comment by jglick** [comment]
Seems to have been abandoned.

**Pull request closed without merging**